task/validate-fbc/0.1/validate-fbc.yaml

```diff
@@ -72,140 +79,133 @@ spec:
         source /utils.sh
         trap 'handle_error "$(results.TEST_OUTPUT.path)"' EXIT
 
-        IMAGE_INSPECT=image_inspect.json
-        BASE_IMAGE_INSPECT=base_image_inspect.json
-        RAW_IMAGE_INSPECT=raw_image_inspect.json
-
-        IMAGE_URL="${IMAGE_URL}@${IMAGE_DIGEST}"
         # Given a tag and a the digest in the IMAGE_URL we opt to use the digest alone
         # this is because containers/image currently doesn't support image references
         # that contain both. See https://github.com/containers/image/issues/1736
-        if [[ "${IMAGE_URL}" == *":"*"@"* ]]; then
-          IMAGE_URL="${IMAGE_URL/:*@/@}"
+        IMAGE_URL="$(get_image_registry_and_repository "${IMAGE_URL}")"
+        image_with_digest="$(get_image_registry_repository_digest "${IMAGE_URL}@${IMAGE_DIGEST}")"
+
+        status=0
+        manifest_digests="$(get_image_manifests -i "$image_with_digest")" || status=$?
+        if [ "$status" -ne 0 ]; then
+            echo "Failed to get manifests for ${image_with_digest}"
+            note="Step inspect-image failed: Encountered errors while inspecting image. For details, check Tekton task log."
+            TEST_OUTPUT=$(make_result_json -r ERROR -t "$note")
+            echo "${TEST_OUTPUT}" | tee "$(results.TEST_OUTPUT.path)"
+            exit 0
+        fi
+        echo "$manifest_digests"
+
+        # Use the amd64 manifest digest for future operations
+        reference_manifest_digest=$(echo -n "${manifest_digests}" | jq -jr '.amd64')
+        if [[ "${reference_manifest_digest}" == "null" ]]; then
+            echo "No amd64 image manifest found for ${image_with_digest}"
+            note="Step inspect-image failed: Encountered errors while inspecting image. For details, check Tekton task log."
+            TEST_OUTPUT=$(make_result_json -r ERROR -t "$note")
+            echo "${TEST_OUTPUT}" | tee "$(results.TEST_OUTPUT.path)"
+            exit 0
+        fi
+        image_with_digest="$(get_image_registry_repository_digest "${IMAGE_URL}@${reference_manifest_digest}")"
+        echo -n "${image_with_digest}" > /shared/image_with_digest
+
+        # FBC content should be identical across different architectures so the IMAGES_PROCESSED result
+        # is a no-op for now. We can come back later and enforce that the content is identical between
+        # all architectures if desired.
+        images_processed_template='{"image": {"pullspec": "'"$IMAGE_URL"'", "digests": [%s]}}'
+        if [ -n "$manifest_digests" ]; then
+          while read -r _arch arch_sha; do
+            digests_processed+=("\"$arch_sha\"")
+          done < <(echo "$manifest_digests" | jq -r 'to_entries[] | "\(.key) \(.value)"')
         fi
 
+        # If the image is an Image Index, also add the Image Index digest to the list.
```

Comment:
Can I suggest moving lines 121 - 125 towards the end of the task? I think we should only emit IMAGES_PROCESSED result once we have successfully finished processing the image. Current approach isn't incorrect either. I just feel if the task fails on line 134, we didn't really process the image. I'll leave the final call on you.

P.S: For error/failure use cases, we can skip emitting the IMAGES_PROCESSED result. EC only verifies the presence of digests if the task actually emits the result.

Reply:
If you want to wait until we process the image then we should actually generate this result in the next step. Since we were already processing the manifests here, I thought that it was best to just do all of the manifest operations in one place (I didn't want to duplicate the manifest operations in two steps).

Since the IMAGES_PROCESSED result doesn't adversely affect policy evaluation, I thought it was better to just get the result generation over with (especially since we are not _really_ processing all of the digests).

Reply:
Well, you could just process everything here and store everything in a file. Then, at the end you could simply emit the result.

 But, like I said, the current approach is correct too. I am fine with keeping it as is.